### PR TITLE
Ensure torch.save() deterministic output

### LINKED ISSUE #42163: torch.save produces inconsistent results between multiple runs
## 🐛 Bug

`torch.save` produces inconsistent results between multiple runs.

## To Reproduce

Run the following snippet multiple times

```python
import hashlib
import torch


def get_sha256_hash(file: str, chunk_size: int = 4096) -> str:
    hasher = hashlib.sha256()
    with open(file, "rb") as fh:
        for chunk in iter(lambda: fh.read(chunk_size), b""):
            hasher.update(chunk)
    return hasher.hexdigest()


file = "tensor.pt"
hashes = []
for _ in range(5):
    obj = torch.ones(1)
    torch.save(obj, file)
    hashes.append(get_sha256_hash(file)[:8])
    del obj

hash = hashes[0]
assert all(other == hash for other in hashes[1:])
print(hash)
```

Aggregated output:

```
4f8ac728
4016cf97
e398369e
```

## Expected behavior

Consistent hash between runs. 

## Environment

```
PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.5.1
[pip3] torchvision==0.6.1
[conda] Could not collect
```

## Additional context

If you replace `obj = torch.ones(1)` with `obj = np.ones(1)` (don't forget to `import numpy as np`) the output is consistently `3cf7a2d4`. 

Cc @albanD

### PR DESCRIPTION
**Pull request by FrancescoCasalegno**
Fixes #42163.

## 🔥 Pitch 

Currently, the binary outputs produced by `torch.save()` are non-deterministic (as pointed out in #42163). This means that running a simple snippet that creates a tensor (or a model) twice will produce output files with a different `md5` sum. 

**Why does this occur?**
The cause of this behavior lies in the fact that the `obj._cdata` is used to identify a tensor and is written to a file, but the `_cdata` attribute is of course non-deterministic:
https://github.com/pytorch/pytorch/blob/a80b215a9ac089cdd3586060467615fa3c4bffe2/torch/serialization.py#L416

**Why does this matter?**
Reproducibility is essential for many Machine Learning projects.
For instance, when using [`dvc`](https://dvc.org/) you would expect that if none of the dependencies of a stage  of a ML pipeline has changed, then running the same stage another time will produce the same binary output. For the reasons explained above, with `torch` this was not the case, so this PR tries to fix this issue.

### What changes?
- The `persistent_id()` function now returns a deterministic value, rather than `obj._cdata` (which depends on runtime).
- As a consequence, `torch.save(obj, "output.pt")` produces a deterministic output, i.e. the `md5` hash of `output.pt` is determinstic. See **Test 1** and **Test 2** below.
 
### What does not change?
- If an `obj` contains several tensors that share the same underlying data (e.g. they are views of the same tensor),the `obj_key` returned by `persistent_id()` is still going to be the same for all of them
- As a consequence, serialization optimizes disk storage by storing only necessary tensors, rather than writing one tensor per view. See **Test 3** below.

### Test 1: snipped from https://github.com/pytorch/pytorch/issues/42163
Consider the following `snippet_1.py` (from #42163).
```python
import hashlib
import torch


def get_sha256_hash(file: str, chunk_size: int = 4096) -> str:
    hasher = hashlib.sha256()
    with open(file, "rb") as fh:
        for chunk in iter(lambda: fh.read(chunk_size), b""):
            hasher.update(chunk)
    return hasher.hexdigest()


file = "tensor.pt"
hashes = []
for _ in range(5):
    obj = torch.ones(1)
    torch.save(obj, file)
    hashes.append(get_sha256_hash(file)[:8])
    del obj

hash = hashes[0]
assert all(other == hash for other in hashes[1:])
print(hash)
```

On `master` you obtain an error
```bash
$ python snippet_1.py
Traceback (most recent call last):
  File "save_tensor.py", line 84, in <module>
    assert all(other == hash for other in hashes[1:])
AssertionError
```
while on this PR branch you should get the following consistent behaviour:
```bash
$ for run in {1..2}; do python snippet_1.py; done
600a83cb
600a83cb
```

### Test 2: Deterministic save of `Tensor` and `nn.Module` instances
Consider the following `snippet_2.py`
```python
import torch
torch.manual_seed(0)
x = torch.tensor([8., 8., 5., 0.])
torch.save(x, "out_tensor.pt")

model = torch.nn.Sequential(
    torch.nn.Linear(3, 1),
    torch.nn.Flatten(0, 1)
)
torch.save(model, "out_model.pt")
```
On `master` branch, the `md5` hash of `out_tensor.pt` and `out_model.pt` are non-determinstic, for instance you may get
```bash
$ for run in {1..2}; do python snippet_2.py; md5 out_*pt; done 
MD5 (out_model.pt) = 92dca4a310b691e893f3cb41d64d5af1
MD5 (out_tensor.pt) = a4ef290583f50a9c203a42d0cfc078af
MD5 (out_model.pt) = de3cb9791a66af8aed77ed7224bd1d5c
MD5 (out_tensor.pt) = 3b8a6009d3a0be5b9dd94152dcc0c7cb
```
while on this PR branch you should get the following consistent behaviour:
```bash
$ for run in {1..2}; do python snippet_2.py; md5 out_*pt; done 
MD5 (out_model.pt) = dba75fd50a190e4e7fa89b7a2477bab7
MD5 (out_tensor.pt) = 029f52f0706d6c813cc796d3cdcd3eb0
MD5 (out_model.pt) = dba75fd50a190e4e7fa89b7a2477bab7
MD5 (out_tensor.pt) = 029f52f0706d6c813cc796d3cdcd3eb0
```

### Test 3: Views of the same tensor are not re-written to file
Consider the following `snippet_3.py`.
```python
import torch
torch.manual_seed(0)
x = torch.rand(1_000, 1_000)
y = x.T
z = x.view(1_000_000, 1)

torch.save({"x": x}, "out_tensor_x.pt")
torch.save({"x": x, "y": y, "z": z}, "out_tensor_xyz.pt")
```
Both on `master` branch and on this  PR branch you should get two output files with same size:
```bash
$ python snippet_3.py && du -sh out_tensor*pt && md5 out_*pt
3.8M    out_tensor_x.pt
3.8M    out_tensor_xyz.pt
MD5 (out_tensor_x.pt) = eda516d9156177b27bdc2a75c9064d9b
MD5 (out_tensor_xyz.pt) = 333b869f5b93ced7b8649ab1571eb8e3
```

### CODE REVIEW

File: torch/serialization.py
Comment (pmeier) [inline torch/serialization.py:470]: You can make this more concise with the [`setdefault`](https://docs.python.org/3.6/library/stdtypes.html#dict.setdefault) method:

```suggestion
            obj_key = persistent_id.id_map.setdefault(obj._cdata, str(len(persistent_id.id_map)))
```

File: torch/serialization.py
Comment (pmeier) [inline torch/serialization.py:480]: Not sure why you bind this to function. Could you explain? IMO it makes it less readable. If there is no "good" reason, I would simply add it as local variable below `serialized_storages` (L458).

**Review by pmeier — APPROVED** [review]
Thanks a lot @FrancescoCasalegno for the fix! I have two minor comments.

File: torch/serialization.py
Comment (FrancescoCasalegno) [reply torch/serialization.py:470]: Nice observation, thanks!

File: torch/serialization.py
Comment (FrancescoCasalegno) [reply torch/serialization.py:478]: The reason was that I thought of `id_map` as a C-style static variable bound to a function (e.g. see [here](https://www.geeksforgeeks.org/static-variables-in-c/)).

Logically, `id_map` is only used by the inner function `persistent_id()` and nowhere in `_save()` we really need to access or even care about the existence of `id_map`. On the contrary, `serialized_storages` is modified by `persistent_id()` but it is also used by `_save()`, so I think that is a slightly different case.

Anyway, in terms of output the two approaches give the same result so if you prefer to move the variable to the outer scope I will do it, no problem!

Comment (FrancescoCasalegno) [reply torch/serialization.py]: Fixed in 1e4cfc7386.

**Comment by FrancescoCasalegno** [comment]
Hello @pmeier  can you please tell me if there's any action still required from my side?

I see that a few tests seem to be failing but I am not sure if there's really something wrong on my side.
1. The only test failing on b12d6e9 was [`render_test_results `](https://github.com/pytorch/pytorch/pull/57536/checks?check_run_id=2509693849) but the error seems to be related to Junit rendering and I am not sure if I can do anything on my side.
2. I re-triggered the tests on 7aaf2eb (empty commit) and then [`pytorch_linux_xenial_py3_6_gcc5_4_build
`](https://app.circleci.com/pipelines/github/pytorch/pytorch/317350/workflows/362d399a-358c-4f83-9df6-4bebf352f8e5/jobs/13214315) failed. There seem to be conflicts on merging but I don't really understand why since the conflicts are on files I did not even touch. 

Do you have any hint or suggestions on how to proceed?

Also, is there still need for another approval before merging (e.g. @ailzhang)?

Thanks in advance!

**Comment by pmeier** [comment]
Hey @FrancescoCasalegno 

> Do you have any hint or suggestions on how to proceed?

Yes. The CI build can be a little opaque at times, so I would suggest you simply rebase your branch on top of `master` or merge it into yours. Usually this should fix the CI even if you don't experience any merge conflicts.

> Also, is there still need for another approval before merging (e.g. @ailzhang)?

Yes. I'm not a FB employee, so we need to wait for one to stamp and merge this PR. If there is no action until tomorrow, I'll ask around who else might be able to review this.

**Review by ailzhang — APPROVED** [review]
LGTM! Thanks for the fix!

**Pull request closed without merging**
